### PR TITLE
feat: temporary slippage

### PR DESCRIPTION
**Pull request by nuanyang233**
- feat: add price updated state
- feat: confirm price impact too high

## Description

Closes #MASK-393

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

**Comment by nuanyang233** [comment]
Wait for test @DimensionDev/qa

**Comment by wenluomask** [comment]
Jira: https://mask.atlassian.net/browse/MASK-393

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx [hunk 1]
```diff
@@ -130,6 +154,20 @@ export function ConfirmDialogUI(props: ConfirmDialogUIProps) {
...
+        cacheTrade.current = null
+    }, [open])
+
+    useUpdateEffect(() => {
+        if (!open) return
+        if (!cacheTrade.current) cacheTrade.current = trade
+        else if (cacheTrade.current?.outputAmount !== trade.outputAmount) {
+            setPriceUpdated(true)
+        }
+    }, [open, trade])
```
Comment (hanszhoou) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx]:
```suggestion
    useUpdateEffect(() => {
        if (!open) {
            setPriceUpdated(false)
            cacheTrade.current = null
            return
        }
        if (!cacheTrade.current) cacheTrade.current = trade
        else if (cacheTrade.current?.outputAmount !== trade.outputAmount) {
            setPriceUpdated(true)
        }
    }, [open, trade])
```

Is one hook enough?

File: packages/icons/general/Cram.tsx [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import { createIcon } from '../utils'
+
+export const CramIcon = createIcon(
+    'CramIcon',
+    <g fill="none">
+        <path opacity=".2" d="M12 2C6.477 2 2 6.477 2 12s4.477 10 10 10 10-4.477 10-10S17.523 2 12 2Z" fill="#F4212E" />
+        <path d="m13.637 6.644-4.633 4.633 2.453 1.635L10.367 17 15 12.367l-2.453-1.635 1.09-4.088Z" fill="#F4212E" />
```
Comment (hanszhoou) [inline packages/icons/general/Cram.tsx]: It's ok to replace #F4212E with redMain(#FF5F5F). I think this is a small design mistake.

File: packages/icons/general/TriangleWarning.tsx [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { createIcon } from '../utils'
+
+export const TriangleWarning = createIcon(
+    'TriangleWarningIcon',
+    <g fill="none">
+        <path opacity=".2" d="m10 .84 10 17.32H0L10 .84Z" fill="#FF5F5F" />
```
Comment (hanszhoou) [inline packages/icons/general/TriangleWarning.tsx:6]: `fill={MaskColorVar.redMain.alpha(1)}`

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx [hunk 4]
```diff
@@ -128,7 +161,53 @@ export function ConfirmDialogUI(props: ConfirmDialogUIProps) {
...
+    }, [currentSlippage])
+
+    const onAccept = useCallback(() => {
+        setPriceUpdated(false)
+        setCacheTrade(trade)
+    }, [trade])
+
+    const onConfirmPriceImpact = useCallback(() => {
+        if (!cacheTrade?.priceImpact) return
+        setTemporarySlippage(new BigNumber(cacheTrade?.priceImpact.multipliedBy(10000).toFixed(0)).toNumber())
```
Comment (hanszhoou) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx]: `const UNISWAP_PERCENT_DENOMINATOR = 10000`

File: packages/icons/general/TriangleWarning.tsx (hunk 3 above)
Comment (nuanyang233) [reply packages/icons/general/TriangleWarning.tsx:6]: `MaskColorVar` cann't be used in the package

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx (hunk 4 above)
Comment (nuanyang233) [reply packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx]: Replace with `PERCENT_DENOMINATOR` to avoid name ambiguity

**Review by hanszhoou — DISMISSED** [review]

**Comment by wenluomask** [comment]
Fixed https://mask.atlassian.net/browse/MASK-417 and https://mask.atlassian.net/browse/MASK-393 .
Checked.

**Review by hanszhoou — DISMISSED** [review]

File: packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx [hunk 5]
```diff
@@ -128,7 +159,54 @@ export function ConfirmDialogUI(props: ConfirmDialogUIProps) {
     )
     // #endregion
 
-    const staled = !!(executionPrice && !executionPrice.isEqualTo(trade.executionPrice))
+    const staled = !!(executionPrice && !executionPrice.isEqualTo(cacheTrade?.executionPrice ?? 0))
+
+    const isGreatThanSlippageSetting = useGreatThanSlippageSetting(cacheTrade?.priceImpact)
+
+    const alertTip = useMemo(() => {
+        if (currentSlippage >= 150 && currentSlippage < 1000) return null
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/SNSAdaptor/trader/ConfirmDialog.tsx]: Refactor those dead code numbers as constants.

**Review by guanbinrui — DISMISSED** [review]

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**